File: 1014.best-sightseeing-pair.py
#
# @lc app=leetcode id=1014 lang=python3
#
# [1014] Best Sightseeing Pair
#

# @lc code=start
class Solution:
    def maxScoreSightseeingPair(self, values: list[int]) -> int:
        preMax = values[0] + 0
        ans = 0
        for i, c in enumerate(values[1:], 1):
            ans = max(ans, preMax + c - i)
            preMax = max(preMax, c + i)
        return ans
# @lc code=end

File: 1014.best-sightseeing-pair.java
/*
 * @lc app=leetcode id=1014 lang=java
 *
 * [1014] Best Sightseeing Pair
 */
// @lc code=start
class Solution {
    public int maxScoreSightseeingPair(int[] values) {
        int maxVal = 0;
        int ans = Integer.MIN_VALUE;
        for (int i = 0; i < values.length; i++){
            int score = values[i] - i; 
            ans = Math.max(ans, maxVal + score);
            maxVal = Math.max(maxVal, values[i] + i);
        }
        return ans;
    }
}
// @lc code=end

